docs/transpile/custom-transpiler-pass.ipynb

```diff
@@ -0,0 +1,352 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from qiskit.dagcircuit import DAGCircuit\n",
+    "from qiskit.circuit import Gate\n",
+    "from qiskit.circuit.library import CXGate, ECRGate\n",
+    "from qiskit.transpiler import PassManager\n",
+    "from qiskit.transpiler.basepasses import TransformationPass\n",
+    "from qiskit.quantum_info import Operator, pauli_basis\n",
+    "\n",
+    "import numpy as np\n",
+    "\n",
+    "from typing import Iterable, Optional"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class PauliTwirl(TransformationPass):\n",
+    "    \"\"\"Add Pauli twirls to two-qubit gates.\"\"\"\n",
+    "\n",
+    "    def __init__(\n",
+    "        self,\n",
+    "        gates_to_twirl: Optional[Iterable[Gate]] = None,\n",
+    "    ):\n",
+    "        \"\"\"\n",
+    "        Args:\n",
+    "            gates_to_twirl: Names of gates to twirl. The default behavior is to twirl all\n",
+    "                two-qubit basis gates, `cx` and `ecr` for IBM backends.\n",
+    "        \"\"\"\n",
+    "        if gates_to_twirl is None:\n",
+    "            gates_to_twirl = [CXGate(), ECRGate()]\n",
+    "        self.gates_to_twirl = gates_to_twirl\n",
+    "        self.build_twirl_set()\n",
+    "        super().__init__()\n",
+    "\n",
+    "    def build_twirl_set(self):\n",
+    "        \"\"\"\n",
+    "        Build set of Paulis to twirl for each gate and store internally as .twirl_set.\n",
+    "        \"\"\"\n",
+    "        self.twirl_set = {}\n",
+    "\n",
+    "        # iterate through gates to be twirled\n",
+    "        for twirl_gate in self.gates_to_twirl:\n",
+    "            twirl_list = []\n",
+    "\n",
+    "            # iterate through Paulis on left of gate to twirl\n",
+    "            for pauli_left in pauli_basis(2):\n",
+    "\n",
+    "                # iterature through Paulis on right of gate to twirl\n",
+    "                for pauli_right in pauli_basis(2):\n",
+    "\n",
+    "                    # save pairs that produce identical operation as gate to twirl\n",
+    "                    if (Operator(pauli_left) @ Operator(twirl_gate)).equiv(Operator(twirl_gate) @ pauli_right):\n",
+    "                        twirl_list.append((pauli_left, pauli_right))\n",
+    "\n",
+    "            self.twirl_set[twirl_gate.name] = twirl_list\n",
+    "\n",
+    "    def run(\n",
+    "        self,\n",
+    "        dag: DAGCircuit,\n",
+    "    ) -> DAGCircuit:\n",
+    "        \n",
+    "        # iterate through gates to be twirled\n",
+    "        for twirl_gate in self.gates_to_twirl:\n",
+    "\n",
+    "            # collect all nodes in DAG with specified name\n",
+    "            for run in dag.collect_runs(twirl_gate.name):\n",
+    "                for node in run:\n",
```

Comment:
In this case you're just iterating over all instances of a specific gate you can just use `DAGCircuit.op_nodes` this will be more efficient as it just iterates over all the nodes in insertion order and yields each operation. The `collect_runs()` method is doing a more involved search to find all the uninterrupted sequences of gates.

Reply:
But I think in general then I would need to compare every instruction to one that I want to twirl, so in general I think this is better (although not for this particular circuit).

Reply:
You actually are already doing that it's just abstracted a bit away as part of the search that rustworkx performs via the dagcircuit api. What the code is doing right now has a lot of overhead, essentially all you're doing is searching for all the instances of `self.gates_to_twirl` in the circuit. Basically you're doing 1 topological sort for each gate and then an iteration over that sorted node list that constructs the list of runs you iterate over again here. If you're curious the `collect_runs()` method here calls this function in rustworkx: https://github.com/Qiskit/rustworkx/blob/main/src/dag_algo/mod.rs#L468-L514 and `filter_fn` arg is doing the gate name check. 

To simplify this you can basically just rewrite the run method as:

```python
        twirling_gate_classes = tuple(gate.base_class for gate in gates_to_twirl)
        for node in dag.op_nodes():
            if not isinstance(node.op, twirling_gate_classes):
                continue
            # random integer to select Pauli twirl pair
            pidx = np.random.randint(0, len(self.twirl_set[twirl_gate.name]),)
            twirl_pair = self.twirl_set[node.op.name][pidx]

            # instantiate mini_dag and attach quantum register
            mini_dag = DAGCircuit()
            register = QuantumRegister(2)
            mini_dag.add_qreg(register)
            # apply left Pauli, gate to twirl, and right Pauli to empty mini-DAG
            mini_dag.apply_operation_back(twirl_pair[0].to_instruction(), [register[0], register[1]])
            mini_dag.apply_operation_back(node.op, [register[0], register[1]])
            mini_dag.apply_operation_back(twirl_pair[1].to_instruction(), [register[0], register[1]])

            # substitute gate to twirl node with twirling mini-DAG
            dag.substitute_node_with_dag(node, mini_dag)
```

Reply:
Gotcha, thanks!

Reply:
Just committed the change that uses `dag.op_nodes()` instead of `.collect_runs()`.